### PR TITLE
Enable buy and sell for decred and fix wallet bug on ios

### PR DESCRIPTION
**Pull request by ukane-philemon**
Issue Number (if Applicable): Fixes #

# Description

1. This PR fixes a bug where existing decred wallets are not found after the wallet app is restarted on ios.
2. This PR enables buy and sell onramp. Note: sell doesn't work rn (from my end) but there's a buy option for decred.

Builds on: #12 

# Pull Request - Checklist  

- [x] Initial Manual Tests Passed
- [x] Double check modified code and verify it with the feature/task requirements
- [x] Format code
- [x] Look for code duplication
- [x] Clear naming for variables and methods

### CODE REVIEW

**Comment by ukane-philemon** [comment]
Note: I've not been able to make onramper select decred as the default currency.

File: lib/entities/provider_types.dart [hunk 1]
```diff
@@ -80,7 +80,7 @@ class ProvidersHelper {
       case WalletType.none:
       case WalletType.haven:
       case WalletType.decred:
-        return [];
+        return [ProviderType.askEachTime, ProviderType.onramper];
```
Comment (JoeGruffins) [inline lib/entities/provider_types.dart:81]: Does this also change the behavior for `WalletType.none` and `WalletType.haven`? I think you need to move decred under the others with this change.

Comment (ukane-philemon) [reply lib/entities/provider_types.dart:81]: Updated.

File: cw_decred/lib/wallet.dart [hunk 2]
```diff
@@ -416,8 +416,32 @@ abstract class DecredWalletBase extends WalletBase<DecredBalance,
...
+      // NOTE: We are only fetching 5 transactions from
+      // libdcrwallet.listTransactions and we could run into issues where the
+      // wrong tx amount is reported if a user creates a send transaction that
+      // spends to more than 5 output, i.e we fetch the first 2, 3 or 4 as part
+      // of an initial request and account for outputs that have the same txid
+      // below but when a second request is made, the last transaction might
+      // overwrite the first batch of transactions or will be ignored since they
+      // share the same id. This has been handled in
+      // transaction_history.dart#L35-59 which is the only functional call site
+      // of this method as of writing.
```
Comment (JoeGruffins) [inline cw_decred/lib/wallet.dart]: The line numbers could change, could you rewrite without them?

**Comment by JoeGruffins** [comment]
I don't think [44556dd](https://github.com/JoeGruffins/cake_wallet/pull/14/commits/44556dd28d86658cf90511a917b6704b53190d30) isn't the right way to handle this bug.

I think we should just make the key a combination of txid:vout, so like "50a595cae0c57244d42a03725948c41041159ddc4f365171cb9fd8084a8ed04d:0", then they are all unique.

**Comment by ukane-philemon** [comment]
> I think we should just make the key a combination of txid:vout, so like "50a595cae0c57244d42a03725948c41041159ddc4f365171cb9fd8084a8ed04d:0", then they are all unique.

Errm, okay. I'll try your suggestion. However, we still need to set output addresses, I think.

**Comment by JoeGruffins** [comment]
> However, we still need to set output addresses, I think.

I could be wrong but I dont think we will. The problem was having different outputs using the same id. The values are already divided up by dcrwallet and we don't need to worry about adding them together. Really the map is not transactions but UTXO (and spent txo) which is confusing but we should leave the name anyway imo.

**Comment by ukane-philemon** [comment]
Okay.

**Comment by JoeGruffins** [comment]
Thanks this looks correct to me!

Two things I noticed, one is pending transactions are not being confirmed. Another is can you change the libwallet version in the android script as well?

**Comment by ukane-philemon** [comment]
> one is pending transactions are not being confirmed.

They are replaced. 

> Another is can you change the libwallet version in the android script as well?

Sure, missed that.

**Comment by JoeGruffins** [comment]
It seems the ui is not updating because the id is not the same as the map id. Can you not add vout to the transaction info and just make the id the txid:vout? You can make dcrdata work in lib/view_model/transaction_details_view_model.dart with `return 'https://dcrdata.decred.org/tx/${txId.substring(0, 64)}';`

**Review by JoeGruffins — APPROVED** [review]

File: cw_decred/lib/wallet.dart (hunk 2 above)
Comment (ukane-philemon) [reply cw_decred/lib/wallet.dart]: In a new commit yh?

Comment (JoeGruffins) [reply cw_decred/lib/wallet.dart]: Oh sorry, this is old, please ignore.

**Comment by JoeGruffins** [comment]
Changes have been added to the upstream pr so closing.

**Pull request closed without merging**